### PR TITLE
Putting together the vsock fix and tests

### PR DESCRIPTION
**Pull request by alexandruag**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jailer/Cargo.toml [hunk 1]
```diff
@@ -16,3 +16,6 @@ fc_util = { path = "../fc_util" }
 
 [dev-dependencies]
 tempfile = ">=3.0.2"
+
+[features]
+vsock = []
```
Comment (aghecenco) [inline jailer/Cargo.toml]: nit: add newline.

File: jailer/src/env.rs [hunk 2]
```diff
@@ -256,6 +262,25 @@ impl Env {
...
+            // Do the same for /dev/vhost_vsock with (major, minor) = (10, 241).
+            if unsafe {
+                libc::mknod(
+                    dev_vhost_vsock_path.as_ptr(),
+                    libc::S_IFCHR | libc::S_IRUSR | libc::S_IWUSR,
+                    libc::makedev(10, 241),
+                )
+            } < 0
+            {
+                return Err(Error::MknodDevVhostVsock(sys_util::Error::last()));
```
Comment (aghecenco) [inline jailer/src/env.rs]: Can we use `std::io::Error` instead of `sys_util::Error`?

LE: Oh, I see plenty of `sys_util::Error` in the jailer, so this is out of scope in this PR

File: vmm/src/default_syscalls/x86_64.rs [hunk 3]
```diff
@@ -576,6 +433,269 @@ pub fn default_context() -> Result<SeccompFilterContext, Error> {
     )?)
 }
 
+fn create_ioctl_seccomp_rule() -> Result<Vec<SeccompRule>, Error> {
+    // We will be appending to this vector based on whether or not we are building with `vsock` feature.
+    #[allow(unused_mut)]
+    let mut ioctl_rules: Vec<SeccompRule> = vec![
```
Comment (aghecenco) [inline vmm/src/default_syscalls/x86_64.rs]: nit: group these together in a `fn create_common_ioctl_seccomp_rule`

File: tests/integration_tests/build/test_build.py [hunk 4]
```diff
@@ -2,60 +2,47 @@
...
+
+    if build_type == "release":
+        extra_args += "--release "
+
+    # The relative path of the binaries is computed using the build_type
+    # (either release or debug) and if any features are provided also using
+    # the features names.
+    # For example, a default release build with no features will end up in
+    # the relative directory "release", but for a vsock release build the
+    # relative directory will be "release-vsock".
```
Comment (aghecenco) [inline tests/integration_tests/build/test_build.py]: What happens if we parametrize for more than one feature at a time? Would `rel_path` still be valid?

File: tests/framework/microvm.py [hunk 5]
```diff
@@ -43,13 +43,18 @@ def __init__(
...
+        build_feature='',
         monitor_memory=True,
         newpid_cloner_path=None
     ):
         """Set up microVM attributes, paths, and data structures."""
         # Unique identifier for this machine.
         self._microvm_id = microvm_id
 
+        # This is used in tests to identify if the microvm was started
+        # using a vsock build or a default build.
```
Comment (aghecenco) [inline tests/framework/microvm.py]: Same question: what about composing features?

File: tests/integration_tests/functional/test_api.py [hunk 6]
```diff
@@ -3,11 +3,30 @@
 """Tests that ensure the correctness of the Firecracker API."""
 
 import os
+import pytest
 
 import host_tools.drive as drive_tools
 import host_tools.logging as log_tools
 import host_tools.network as net_tools
 
+def test_dummy_test(test_microvm_with_api):
```
Comment (aghecenco) [inline tests/integration_tests/functional/test_api.py]: nit: 2 blank lines between functions

File: tests/integration_tests/build/test_build.py (hunk 4 above)
Comment (andreeaflorescu) [reply tests/integration_tests/build/test_build.py]: The code as is does not support multiple features (mostly because I thought it was very unlikely for us to add any more). But I can implement support for more features as well.

File: vmm/src/lib.rs [hunk 7]
```diff
@@ -1225,8 +1225,11 @@ impl Vmm {
             error!("Failed to log metrics while stopping: {}", e);
         }
 
-        // Exit from Firecracker using the provided exit code.
-        std::process::exit(exit_code);
+        // Exit from Firecracker using the provided exit code. Safe because we're terminating
+        // the process anyway.
+        unsafe {
+            libc::_exit(exit_code);
```
Comment (andreeaflorescu) [inline vmm/src/lib.rs]: But why is it better to terminate the process with libc::exit? Feel free to add the explanation to the commit message as well.

File: vmm/src/default_syscalls/x86_64.rs [hunk 8]
```diff
@@ -576,6 +433,269 @@ pub fn default_context() -> Result<SeccompFilterContext, Error> {
     )?)
 }
 
+fn create_ioctl_seccomp_rule() -> Result<Vec<SeccompRule>, Error> {
```
Comment (andreeaflorescu) [inline vmm/src/default_syscalls/x86_64.rs]: Are these actually ioctls?

File: tests/integration_tests/functional/test_api.py (hunk 6 above)
Comment (andreeaflorescu) [reply tests/integration_tests/functional/test_api.py]: Done

File: tests/integration_tests/functional/test_api.py [hunk 9]
```diff
@@ -743,3 +744,50 @@ def _drive_patch(test_microvm):
...
+
+    test_microvm = test_microvm_with_api
+    test_microvm.spawn()
+    test_microvm.basic_config()
+
+    response = test_microvm.vsock.put(
+        vsock_id='vsock1',
+        guest_cid=15
+    )
+    assert test_microvm.api_session.is_good_response(response.status_code)
```
Comment (andreeaflorescu) [inline tests/integration_tests/functional/test_api.py]: We switched from asserting on is_good_response to the actual returned status code. Please use `is_status_no_content` or whatever is the appropriate method here.

File: tests/integration_tests/functional/test_api.py [hunk 10]
```diff
@@ -743,3 +744,50 @@ def _drive_patch(test_microvm):
...
+
+    # No other vsock action is allowed after booting the VM.
+    test_microvm.start()
+
+    # Updating an existing vsock should not be fine at this point.
+    response = test_microvm.vsock.put(
+        vsock_id='vsock1',
+        guest_cid=17
+    )
+    assert not test_microvm.api_session.is_good_response(response.status_code)
```
Comment (andreeaflorescu) [inline tests/integration_tests/functional/test_api.py]: Use `is_status_bad_request` please.

File: tests/integration_tests/functional/test_vsock.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (andreeaflorescu) [inline tests/integration_tests/functional/test_vsock.py]: 2019*

File: CHANGELOG.md [hunk 12]
```diff
@@ -19,6 +19,8 @@
 
 ### Fixed
 - virtio-blk: VIRTIO_BLK_T_FLUSH now working as expected.
+- The vsock feature can be used when starting Firecracker using the jailer
```
Comment (andreeaflorescu) [inline CHANGELOG.md]: I would split this in 2:
- in the added section: Jailer support for firecracker vsock builds.
- in the fixed section: Updated seccomp whitelist to allow seccomp level = 2 with vsock

File: tests/host_tools/test_vsock.c [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#define SERVER_PORT     1100
+
+static const int msg = 1235;
+
+int print_usage() {
+  printf("Usage: ./vsock_test { client <server_cid> | server }\n");
```
Comment (aghecenco) [inline tests/host_tools/test_vsock.c]: nit: 4 spaces would improve readability

File: tests/integration_tests/functional/test_api.py [hunk 14]
```diff
@@ -5,11 +5,35 @@
...
+def test_dummy_test(test_microvm_with_api):
+    """Dummy test in which we check the build feature."""
+    if test_microvm_with_api.build_feature == 'vsock':
+        print("This is a microvm built with the vsock feature enabled.")
+
+
+def test_dummy_only_for_vsock_build(test_microvm_with_api):
+    """Dummy test that is executed only for vsock builds."""
+    if test_microvm_with_api.build_feature != 'vsock':
+        pytest.skip("This test is meant only for vsock builds")
```
Comment (aghecenco) [inline tests/integration_tests/functional/test_api.py]: nit: add full stop

File: tests/integration_tests/functional/test_api.py [hunk 15]
```diff
@@ -5,11 +5,35 @@
...
+
+    assert test_microvm_with_api.build_feature == 'vsock'
+
+
+def test_dummy_only_default_build(test_microvm_with_api):
+    """Dummy test that is executed only for default builds."""
+    if test_microvm_with_api.build_feature != '':
+        pytest.skip("This test is meant only for default builds")
```
Comment (aghecenco) [inline tests/integration_tests/functional/test_api.py]: nit: add full stop

File: tests/integration_tests/functional/test_api.py [hunk 16]
```diff
@@ -720,3 +744,50 @@ def _drive_patch(test_microvm):
         path_on_host=test_microvm.create_jailed_resource(fs.path)
     )
     assert test_microvm.api_session.is_status_no_content(response.status_code)
+
+
+def test_api_vsock(test_microvm_with_api):
+    """Test vsock related API commands."""
+    if test_microvm_with_api.build_feature != 'vsock':
+        pytest.skip("This test is meant only for vsock builds")
```
Comment (aghecenco) [inline tests/integration_tests/functional/test_api.py]: nit: add full stop
